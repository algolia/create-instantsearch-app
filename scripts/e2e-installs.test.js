const fs = require('fs');
const { execSync } = require('child_process');

describe('Installation', () => {
  let temporaryDirectory;
  let appPath;

  beforeAll(() => {
    temporaryDirectory = execSync(
      'mktemp -d 2>/dev/null || mktemp -d -t "appPath"'
    )
      .toString()
      .trim();
  });

  afterAll(() => {
    execSync(`rm -rf "${temporaryDirectory}"`);
  });

  beforeEach(() => {
    appPath = `${temporaryDirectory}/test-app`;
    execSync(`mkdir ${appPath}`);
  });

  afterEach(() => {
    execSync(`rm -rf "${appPath}"`);
  });

  describe('Dependencies', () => {
    test('get installed by default', () => {
      execSync(
        `yarn start ${appPath} \
          --template "InstantSearch.js"`,
        { stdio: 'ignore' }
      );

      expect(fs.lstatSync(`${appPath}/node_modules`).isDirectory()).toBe(true);
    });

    test('get skipped with the `no-installation` flag', () => {
      execSync(
        `yarn start ${appPath} \
          --template "InstantSearch.js" \
          --no-installation`,
        { stdio: 'ignore' }
      );

      expect(fs.existsSync(`${appPath}/node_modules`)).toBe(false);
    });
  });

  describe('Path', () => {
    test('without conflict generates files', () => {
      execSync(
        `yarn start ${appPath} \
          --template "InstantSearch.js" \
          --no-installation`,
        { stdio: 'ignore' }
      );

      expect(fs.existsSync(`${appPath}/package.json`)).toBe(true);
    });

    test('with conflict with a non-empty folder cancels generation', () => {
      execSync(`echo 'hello' > ${appPath}/README.md`);

      expect(() => {
        execSync(
          `yarn start ${appPath} \
          --template "InstantSearch.js" \
          --no-installation`,
          { stdio: 'ignore' }
        );
      }).toThrow();

      expect(
        execSync(`grep "hello" ${appPath}/README.md`)
          .toString()
          .trim()
      ).toBe('hello');
    });

    test('with conflict with an existing file cancels generation', () => {
      execSync(`touch ${appPath}/file`);

      expect(() => {
        execSync(
          `yarn start ${appPath}/file \
          --template "InstantSearch.js" \
          --no-installation`,
          { stdio: 'ignore' }
        );
      }).toThrow();

      expect(fs.existsSync(`${appPath}/file`)).toBe(true);
    });
  });
});
