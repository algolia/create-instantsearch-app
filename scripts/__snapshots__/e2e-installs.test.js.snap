// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots InstantSearch.js File content: .gitignore 1`] = `
"# See https://help.github.com/ignore-files/ for more about ignoring files.

# dependencies
/node_modules

# testing
/coverage

# production
/dist
/.cache

# misc
.DS_Store
.env.local
.env.development.local
.env.test.local
.env.production.local

npm-debug.log*
yarn-debug.log*
yarn-error.log*"
`;

exports[`Snapshots InstantSearch.js File content: README.md 1`] = `
"# instantsearch.js-app

_This project was generated with [create-instantsearch-app](https://github.com/algolia/create-instantsearch-app) by [Algolia](https://algolia.com)._

## Get started

To run this project locally, install the dependencies and run the local server:

\`\`\`sh
npm install
npm start
\`\`\`

Alternatively, you may use [Yarn](https://http://yarnpkg.com/):

\`\`\`sh
yarn
yarn start
\`\`\`"
`;

exports[`Snapshots InstantSearch.js File content: index.html 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
  <meta charset=\\"utf-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\">
  <meta name=\\"theme-color\\" content=\\"#000000\\">

  <link rel=\\"manifest\\" href=\\"./manifest.webmanifest\\">
  <link rel=\\"shortcut icon\\" href=\\"./favicon.png\\">

  <link rel=\\"stylesheet\\" href=\\"https://cdn.jsdelivr.net/npm/instantsearch.js@1.0.0/dist/instantsearch.min.css\\">
  <link rel=\\"stylesheet\\" href=\\"https://cdn.jsdelivr.net/npm/instantsearch.js@1.0.0/dist/instantsearch-theme-algolia.min.css\\">
  <link rel=\\"stylesheet\\" href=\\"./src/index.css\\">
  <link rel=\\"stylesheet\\" href=\\"./src/app.css\\">

  <title>instantsearch.js-app</title>
</head>

<body>
  <header class=\\"header\\">
    <h1 class=\\"header-title\\">
      <a href=\\"/\\">instantsearch.js-app</a>
    </h1>
    <p class=\\"header-subtitle\\">using InstantSearch.js</p>
  </header>

  <div class=\\"container\\">
    <div class=\\"container-app\\">
      <div class=\\"filters\\">
        <div id=\\"facet1-list\\"></div>
        <div id=\\"facet2-list\\"></div>
      </div>

      <div class=\\"search-container\\">
        <div id=\\"searchBox\\"></div>
        <div id=\\"hits\\"></div>
      </div>
    </div>

    <div id=\\"pagination\\"></div>
  </div>

  <script src=\\"https://cdn.jsdelivr.net/npm/instantsearch.js@1.0.0/dist/instantsearch.min.js\\"></script>
  <script src=\\"./src/app.js\\"></script>
</body>

</html>"
`;

exports[`Snapshots InstantSearch.js File content: manifest.webmanifest 1`] = `
"{
  \\"short_name\\": \\"instantsearch.js-app\\",
  \\"name\\": \\"instantsearch.js-app Sample\\",
  \\"icons\\": [
    {
      \\"src\\": \\"favicon.png\\",
      \\"sizes\\": \\"64x64 32x32 24x24 16x16\\",
      \\"type\\": \\"image/x-icon\\"
    }
  ],
  \\"start_url\\": \\"./index.html\\",
  \\"display\\": \\"standalone\\",
  \\"theme_color\\": \\"#000000\\",
  \\"background_color\\": \\"#ffffff\\"
}"
`;

exports[`Snapshots InstantSearch.js File content: package.json 1`] = `
"{
  \\"name\\": \\"instantsearch.js-app\\",
  \\"version\\": \\"1.0.0\\",
  \\"private\\": true,
  \\"scripts\\": {
    \\"start\\": \\"parcel index.html --port 3000\\",
    \\"build\\": \\"parcel build index.html\\"
  },
  \\"devDependencies\\": {
    \\"parcel-bundler\\": \\"^1.8.1\\"
  }
}"
`;

exports[`Snapshots InstantSearch.js File content: src/app.css 1`] = `
"em {
  background: cyan;
  font-style: normal;
}

.header {
  display: flex;
  align-items: center;
  min-height: 50px;
  padding: 0.5rem 1rem;
  background-image: linear-gradient(284deg, #fedd4e, #fcb43a);
  color: #fff;
  margin-bottom: 1rem;
}

.header-title {
  font-size: 1.2rem;
  font-weight: normal;
}

.header-title::after {
  content: ' ▸ ';
  padding: 0 0.5rem;
}

.header-title a {
  color: #fff;
  text-decoration: none;
}

.header-subtitle {
  font-size: 1.2rem;
}

.container {
  max-width: 1200px;
  margin: 0 auto;
  padding: 1rem;
}

.container-app {
  display: grid;
  grid-template-columns: 20% 75%;
  grid-gap: 5%;
}

.ais-hits {
  display: grid;
  grid-template-columns: 47.5% 47.5%;
  grid-gap: 1rem;
}

.ais-hits--item {
  min-height: 100px;
  padding: 1rem;
  background: #fff;
  border-radius: 4px;
  border: 1px solid rgba(150, 150, 150, 0.16);
  box-shadow: 0 1px 1px 0 rgba(0, 0, 0, 0.16);
}

#searchBox {
  margin-bottom: 2rem;
}

#pagination {
  margin: 2rem auto;
  text-align: center;
}"
`;

exports[`Snapshots InstantSearch.js File content: src/app.js 1`] = `
"/* global instantsearch */

const search = instantsearch({
  appId: 'appId',
  apiKey: 'apiKey',
  indexName: 'indexName',
});

search.addWidget(
  instantsearch.widgets.searchBox({
    container: '#searchBox',
    placeholder: 'Search placeholder',
  })
);

search.addWidget(
  instantsearch.widgets.hits({
    container: '#hits',
    templates: {
      item: '<div class=\\"hit\\">{{{_highlightResult.mainAttribute.value}}}</div>',
    },
  })
);

search.addWidget(
  instantsearch.widgets.refinementList({
    container: '#facet1-list',
    attributeName: 'facet1',
  })
);

search.addWidget(
  instantsearch.widgets.refinementList({
    container: '#facet2-list',
    attributeName: 'facet2',
  })
);


search.addWidget(
  instantsearch.widgets.pagination({
    container: '#pagination',
  })
);

search.start();"
`;

exports[`Snapshots InstantSearch.js File content: src/index.css 1`] = `
"body,
h1 {
  margin: 0;
  padding: 0;
}

body {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica,
    Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';
}"
`;

exports[`Snapshots InstantSearch.js Folder structure: contains the right files 1`] = `
Array [
  ".gitignore",
  "README.md",
  "favicon.png",
  "index.html",
  "manifest.webmanifest",
  "package.json",
  "src/app.css",
  "src/app.js",
  "src/index.css",
]
`;

exports[`Snapshots React InstantSearch File content: .gitignore 1`] = `
"# See https://help.github.com/ignore-files/ for more about ignoring files.

# dependencies
/node_modules

# testing
/coverage

# production
/build

# misc
.DS_Store
.env.local
.env.development.local
.env.test.local
.env.production.local

npm-debug.log*
yarn-debug.log*
yarn-error.log*"
`;

exports[`Snapshots React InstantSearch File content: README.md 1`] = `
"# react-instantsearch-app

_This project was generated with [create-instantsearch-app](https://github.com/algolia/create-instantsearch-app) by [Algolia](https://algolia.com)._

## Get started

To run this project locally, install the dependencies and run the local server:

\`\`\`sh
npm install
npm start
\`\`\`

Alternatively, you may use [Yarn](https://http://yarnpkg.com/):

\`\`\`sh
yarn
yarn start
\`\`\`"
`;

exports[`Snapshots React InstantSearch File content: package.json 1`] = `
"{
  \\"name\\": \\"react-instantsearch-app\\",
  \\"version\\": \\"1.0.0\\",
  \\"private\\": true,
  \\"scripts\\": {
    \\"start\\": \\"react-scripts start\\",
    \\"build\\": \\"react-scripts build\\"
  },
  \\"dependencies\\": {
    \\"react\\": \\"^16.3.2\\",
    \\"react-dom\\": \\"^16.3.2\\",
    \\"react-instantsearch\\": \\"^1.0.0\\",
    \\"react-scripts\\": \\"1.1.4\\"
  },
  \\"devDependencies\\": {
    \\"prop-types\\": \\"^15.6.1\\"
  }
}"
`;

exports[`Snapshots React InstantSearch File content: public/index.html 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">

<head>
  <meta charset=\\"utf-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\">
  <meta name=\\"theme-color\\" content=\\"#000000\\">

  <link rel=\\"manifest\\" href=\\"%PUBLIC_URL%/manifest.json\\">
  <link rel=\\"shortcut icon\\" href=\\"%PUBLIC_URL%/favicon.png\\">

  <!--
      Do not use @7 in production, use a complete version like x.x.x, see website for latest version:
      https://community.algolia.com/react-instantsearch/Getting_started.html#load-the-algolia-theme
    -->
  <link rel=\\"stylesheet\\" href=\\"https://cdn.jsdelivr.net/npm/instantsearch.css@7/themes/algolia-min.css\\">

  <title>react-instantsearch-app</title>
</head>

<body>
  <noscript>
    You need to enable JavaScript to run this app.
  </noscript>

  <div id=\\"root\\"></div>
</body>

</html>"
`;

exports[`Snapshots React InstantSearch File content: public/manifest.json 1`] = `
"{
  \\"short_name\\": \\"react-instantsearch-app\\",
  \\"name\\": \\"react-instantsearch-app Sample\\",
  \\"icons\\": [
    {
      \\"src\\": \\"favicon.png\\",
      \\"sizes\\": \\"64x64 32x32 24x24 16x16\\",
      \\"type\\": \\"image/x-icon\\"
    }
  ],
  \\"start_url\\": \\"./index.html\\",
  \\"display\\": \\"standalone\\",
  \\"theme_color\\": \\"#000000\\",
  \\"background_color\\": \\"#ffffff\\"
}"
`;

exports[`Snapshots React InstantSearch File content: src/App.css 1`] = `
"em {
  background: cyan;
  font-style: normal;
}

.header {
  display: flex;
  align-items: center;
  min-height: 50px;
  padding: 0.5rem 1rem;
  background-image: linear-gradient(to right, #8e43e7, #00aeff);
  color: #fff;
  margin-bottom: 1rem;
}

.header a {
  color: #fff;
  text-decoration: none;
}

.header-title {
  font-size: 1.2rem;
  font-weight: normal;
}

.header-title::after {
  content: ' ▸ ';
  padding: 0 0.5rem;
}

.header-subtitle {
  font-size: 1.2rem;
}

.container {
  max-width: 1200px;
  margin: 0 auto;
  padding: 1rem;
}

.container-app {
  display: grid;
  grid-template-columns: 20% 75%;
  grid-gap: 5%;
}

.searchBox {
  margin-bottom: 2rem;
}

.pagination {
  margin: 2rem auto;
  text-align: center;
}"
`;

exports[`Snapshots React InstantSearch File content: src/App.js 1`] = `
"import React, { Component } from 'react';
import {
  InstantSearch,
  Hits,
  SearchBox,
  RefinementList,
  Pagination,
  Highlight,
} from 'react-instantsearch/dom';
import PropTypes from 'prop-types';
import './App.css';

class App extends Component {
  render() {
    return (
      <div>
        <header className=\\"header\\">
          <h1 className=\\"header-title\\">
            <a href=\\"/\\">react-instantsearch-app</a>
          </h1>
          <p className=\\"header-subtitle\\">
            using{' '}
            <a href=\\"https://github.com/algolia/react-instantsearch\\">
              React InstantSearch
            </a>
          </p>
        </header>

        <div className=\\"container\\">
          <InstantSearch
            appId=\\"appId\\"
            apiKey=\\"apiKey\\"
            indexName=\\"indexName\\"
          >
            <div className=\\"container-app\\">
              <div className=\\"filters\\">
                <RefinementList attribute=\\"facet1\\" />
                <RefinementList attribute=\\"facet2\\" />
              </div>

              <div className=\\"search-container\\">
                <SearchBox className=\\"searchBox\\" placeholder=\\"Search placeholder\\" />
                <Hits hitComponent={Hit} />

                <div className=\\"pagination\\">
                  <Pagination />
                </div>
              </div>
            </div>
          </InstantSearch>
        </div>
      </div>
    );
  }
}

function Hit(props) {
  return (
    <div>
      <Highlight attribute=\\"mainAttribute\\" hit={props.hit} />
    </div>
  );
}

Hit.propTypes = {
  hit: PropTypes.object.isRequired,
};

export default App;"
`;

exports[`Snapshots React InstantSearch File content: src/index.css 1`] = `
"body,
h1 {
  margin: 0;
  padding: 0;
}

body {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica,
    Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';
}"
`;

exports[`Snapshots React InstantSearch File content: src/index.js 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
import App from './App';

ReactDOM.render(<App />, document.getElementById('root'));"
`;

exports[`Snapshots React InstantSearch Folder structure: contains the right files 1`] = `
Array [
  ".gitignore",
  "README.md",
  "package.json",
  "public/favicon.png",
  "public/index.html",
  "public/manifest.json",
  "src/App.css",
  "src/App.js",
  "src/index.css",
  "src/index.js",
]
`;
